File: codes/java/124.二叉树中的最大路径和.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
	private Integer ans = Integer.MIN_VALUE;

	public int maxPathSum(TreeNode root) {
		this.ans = Integer.MIN_VALUE;
		dfs(root);
		return this.ans;
	}

	private int dfs(TreeNode root) {
		if (root == null) {
			return 0;
		}
		int left = Math.max(0, dfs(root.left));
		int right = Math.max(0, dfs(root.right));
		this.ans = Math.max(this.ans, root.val + left + right);
		return root.val + Math.max(left, right);
	}
}

File: codes/python/124.二叉树中的最大路径和.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = float('-inf')
        self.dfs(root)
        return self.ans


    def dfs(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        left = max(0, self.dfs(root.left))
        right = max(0, self.dfs(root.right))
        self.ans = max(self.ans, root.val + left + right)
        return root.val + max(left, right)
        